src/platform/webusb/transfer.rs

```diff
@@ -0,0 +1,189 @@
+            data: vec![],
+        }
+    }
+}
+
+impl TransferData {
+    fn store_and_notify(
+        data: Vec<u8>,
+        status: UsbTransferStatus,
+        written_bytes: usize,
+        user_data: *mut c_void,
+    ) {
+        unsafe {
+            let transfer = user_data as *mut TransferInner<TransferData>;
+            let t = (*transfer).platform_data();
+
+            t.data = data;
+            t.status = status;
+            t.written_bytes = written_bytes;
+            notify_completion::<TransferData>(user_data);
+        }
+    }
+}
+
+impl PlatformTransfer for TransferData {
+    fn cancel(&self) {
+        todo!()
+    }
+}
+
+impl PlatformSubmit<Vec<u8>> for TransferData {
+    unsafe fn submit(&mut self, data: Vec<u8>, user_data: *mut c_void) {
+        let device = self.device.clone();
+        let ep_type = self.ep_type;
+        let endpoint_number = self.endpoint;
+        spawn_local(async move {
+            let (written_bytes, status) = match ep_type {
+                EndpointType::Control => {
+                    panic!("Control is unsupported for submit");
+                }
+                EndpointType::Isochronous => {
+                    panic!("Isochronous is unsupported for submit");
+                }
+                EndpointType::Bulk => {
+                    let array = Uint8Array::from(data.as_slice());
+                    let array_obj = Object::try_from(&array).unwrap();
+
+                    let result = JsFuture::from(
+                        device
+                            .device
+                            .transfer_out_with_buffer_source(endpoint_number, array_obj)
+                            .unwrap(),
+                    )
+                    .await
+                    .unwrap();
+
+                    let transfer_result: UsbOutTransferResult = JsCast::unchecked_from_js(result);
+                    (
+                        transfer_result.bytes_written() as usize,
+                        transfer_result.status(),
+                    )
+                }
+                EndpointType::Interrupt => {
+                    panic!("Interrupt is unsupported for submit");
+                }
+            };
+
+            Self::store_and_notify(data, status, written_bytes, user_data);
+        });
+    }
+
+    unsafe fn take_completed(&mut self) -> Completion<ResponseBuffer> {
+        let data = ResponseBuffer::from_vec(self.data.clone(), self.written_bytes);
+        let status = self.status;
+        tracing::info!("{status:?}");
+        Completion {
+            data,
+            status: Ok(()),
+        }
+    }
+}
+
+impl PlatformSubmit<RequestBuffer> for TransferData {
+    unsafe fn submit(&mut self, data: RequestBuffer, user_data: *mut c_void) {
+        let device = self.device.clone();
+        let ep_type = self.ep_type;
+        let endpoint_number = self.endpoint & (!0x80);
+        let (mut data, len) = data.into_vec();
+        spawn_local(async move {
```

Comment:
I need to look at how Rust async and the JS event loop interoperate, but turning the WebUSB promise into a future and spawning a task to await that future seems like extra steps. I think it should be possible to make a JS closure wrapping a Rust function that that copies the data into linear memory, marks the transfer as complete, and notifies the transfer's waker, basically in the place of the event thread that does that on native platforms.

Reply:
You will have to go through the future though to make this transfer happen. I am not sure which layer exactly you want to shed?

Reply:
So `spawn_local` creates a task with a waker, and `JsFuture::from` [creates a JS callback](https://docs.rs/wasm-bindgen-futures/0.4.49/src/wasm_bindgen_futures/lib.rs.html#138-176) that arranges for that waker to be woken when the JS promise completes. That waker then queues a poll of the task, where your code handles the transfer completion, and calls `notify_completion`. The common code wakes a waker stored on the transfer (the waker of the task polling the future returned by nusb), which then queues the user's task.

The TransferInner / PlatformTransfer internals are kind of a mismatch here, because other nusb platforms implement their own event loop but JS comes with one. But if you ignore the event loop itself then `notify_completion` is effectively a completion callback invoked by "platform code". You could use `promise.then()` directly with a callback that calls `notify_completion` instead of converting the promise to a `JsFuture`. One wrinkle is that it needs both success and error callbacks, where `JsFuture` uses `Rc` to share state between the two.

Alternatively, I wonder if there would be a clean way to make the nusb returned future somehow poll the JsFuture, instead of reimplementing what JsFuture does, but I think that would be an invasive change to the cross-platform code.

Not essential and not even sure how much of a performance hit it is to have an extra `queueMicrotask` hop back to JS, but just a thought.

Reply:
Ah, yeah, I did not consider the double spawn/waker. I will fix that :) it's bad enough to have so many microticks :D